src/analysis/refactor_extract_region.ml

```diff
@@ -0,0 +1,395 @@
+open Std
+
+exception Nothing_to_do
+exception Not_allowed_in_interface_file
+
+let () =
+  Location.register_error_of_exn (function
+    | Nothing_to_do -> Some (Location.error "Nothing to do")
+    | Not_allowed_in_interface_file ->
+      Some
+        (Location.error
+           "Expression extraction is only allowed in implementation file")
+    | _ -> None)
+
+module Fresh_name = struct
+  (* Generate a fresh name that does not already exist in given environment. *)
+  let gen_val_name basename env =
+    let rec loop n =
+      let guess = basename ^ Int.to_string n in
+      if Env.bound_value guess env then succ n |> loop else guess
+    in
+    loop 1
+end
```

Comment:
We have a lot of discussion related to identifying free variables, but can we not just using `Env.bound_value`? For example: 

```ocaml
let a = 10 
let b = 
  let c = 2 in 
  let d = 3 in 
  let z = a + c d 
  in z + 1
``

Why not using the `env` related to `b` ?